src/libstd/fs.rs

```diff
@@ -1717,13 +1747,83 @@ mod tests {
         check!(fs::create_dir_all(&dtt));
         check!(fs::create_dir_all(&d2));
         check!(check!(File::create(&canary)).write(b"foo"));
-        check!(fs::soft_link(&d2, &dt.join("d2")));
+        check!(symlink(&d2, &dt.join("d2")));
         check!(fs::remove_dir_all(&d1));
 
         assert!(!d1.is_dir());
         assert!(canary.exists());
     }
 
+    #[cfg(not(windows))]
+    #[test]
+    fn recursive_rmdir_of_symlink() {
+        use os::unix::fs::symlink;
+        let tmpdir = tmpdir();
+        let d1 = tmpdir.join("d1");
+        let d2 = tmpdir.join("d2");
+        let canary = d2.join("do_not_delete");
+        check!(fs::create_dir_all(&d2));
+        check!(check!(File::create(&canary)).write(b"foo"));
+        check!(symlink(&d2, &d1));
+        check!(fs::remove_dir_all(&d1));
+
+        assert!(!d1.is_dir());
+        assert!(canary.exists());
+    }
+
+    #[cfg(windows)]
+    #[test]
+    fn recursive_rmdir() {
+        use os::windows::fs::{symlink_file, symlink_dir};
+        let tmpdir = tmpdir();
+        let d1 = tmpdir.join("d1");
+        let dt = d1.join("t");
+        let dtt = dt.join("t");
+        let d2 = tmpdir.join("d2");
+        let canary = d2.join("do_not_delete");
+        check!(fs::create_dir_all(&dtt));
+        check!(fs::create_dir_all(&d2));
+        check!(check!(File::create(&canary)).write(b"foo"));
+        check!(symlink_dir(&d2, &dt.join("d2")));
+        check!(symlink_file(&canary, &d1.join("canary")));
+        check!(fs::remove_dir_all(&d1));
+
+        assert!(!d1.is_dir());
+        assert!(canary.exists());
+    }
+
+    #[cfg(windows)]
+    #[test]
+    fn recursive_rmdir_of_symlink() {
+        use os::windows::fs::symlink_dir;
+        let tmpdir = tmpdir();
+        let d1 = tmpdir.join("d1");
+        let d2 = tmpdir.join("d2");
+        let canary = d2.join("do_not_delete");
+        check!(fs::create_dir_all(&d2));
+        check!(check!(File::create(&canary)).write(b"foo"));
+        check!(symlink_dir(&d2, &d1));
+        check!(fs::remove_dir_all(&d1));
+
+        assert!(!d1.is_dir());
+        assert!(canary.exists());
+    }
+
+    #[cfg(windows)]
```

Comment:
I think the [`PrivilegeCheck`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa379304%28v=vs.85%29.aspx) API looks like it may do the trick? (I haven't tested this out though).